File: app/src/main/java/com/quigglesproductions/secureimageviewer/room/databases/unified/entity/relations/RoomUnifiedEmbeddedCategory.java
package com.quigglesproductions.secureimageviewer.room.databases.unified.entity.relations;

import android.content.Context;

import androidx.room.Embedded;
import androidx.room.Ignore;
import androidx.room.Junction;
import androidx.room.Relation;

import com.quigglesproductions.secureimageviewer.SortType;
import com.quigglesproductions.secureimageviewer.datasource.folder.IFolderDataSource;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDatabaseFolder;
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDisplayFolder;
import com.quigglesproductions.secureimageviewer.room.databases.unified.UnifiedFileDatabase;
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedCategory;
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedFile;
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedFileCategoryCrossRef;
import com.quigglesproductions.secureimageviewer.ui.enhancedfolderviewer.FolderOrigin;

import java.io.File;
import java.io.FileNotFoundException;
import java.net.URL;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class RoomUnifiedEmbeddedCategory implements IDisplayFolder, IDatabaseFolder {
    @Embedded
    public RoomUnifiedCategory category;
    @Relation(parentColumn = "CategoryId",entityColumn = "FileId",associateBy = @Junction(RoomUnifiedFileCategoryCrossRef.class),entity = RoomUnifiedFile.class)
    public List<RoomUnifiedEmbeddedFile> files;
    //@Relation(parentColumn = "OnlineThumbnailId",entityColumn = "OnlineId",entity = RoomDatabaseFile.class)
    @Ignore
    public RoomUnifiedEmbeddedFile thumbnailFile;
    @Ignore
    IFolderDataSource dataSource;

    public RoomUnifiedEmbeddedCategory(){

    }
    @Override
    public IFolderDataSource getDataSource() {
        /*if(dataSource == null)
            dataSource = new IFolderDataSource() {
                @Override
                public URL getFolderURL() {
                    return null;
                }

                @Override
                public void getFilesFromDataSource(Context context, FolderDataSourceCallback callback, SortType sortType) {
                    callback.FolderFilesRetrieved(files.stream().map(x->(IDisplayFile)x).collect(Collectors.toList()), null);
                }

                @Override
                public void getThumbnailFromDataSource(Context context, UnifiedFileDatabase database, FolderDataSourceCallback callback) {
                    if(thumbnailFile == null && files.size()>0)
                        callback.FolderThumbnailRetrieved(files.get(0).getThumbnailFile(),null);
                    else
                        callback.FolderThumbnailRetrieved(null, new FileNotFoundException());
                }
            };*/
        return dataSource;
    }

    @Override
    public boolean hasUpdates() {
        return false;
    }

    @Override
    public String getName() {
        return category.getName();
    }

    @Override
    public void setHasUpdates(boolean b) {

    }

    public long getOnlineId() {
        return category.getOnlineId();
    }

    @Override
    public void setDataSource(IFolderDataSource retrofitFolderDataSource) {
        dataSource = retrofitFolderDataSource;
    }

    @Override
    public FolderOrigin getFolderOrigin() {
        return FolderOrigin.ROOM;
    }

    @Override
    public void sortFiles(SortType newSortType) {
        switch (newSortType){
            case NAME_ASC:
                files.sort(Comparator.comparing(IDisplayFile::getName));
                break;
            case NAME_DESC:
                files.sort(Comparator.comparing(IDisplayFile::getName).reversed());
                break;
            case NEWEST_FIRST:
                files.sort(Comparator.comparing(IDisplayFile::getDefaultSortTime).reversed());
                break;
            case OLDEST_FIRST:
                files.sort(Comparator.comparing(IDisplayFile::getDefaultSortTime));
                break;
            default:
                files.sort(Comparator.comparing(IDisplayFile::getName));
                break;
        }
    }

    @Override
    public List<IDisplayFile> getFiles() {
        return files.stream().map(x->(IDisplayFile)x).collect(Collectors.toList());
    }

    @Override
    public long getId() {
        return category.getUid();
    }

    @Override
    public File getThumbnailFile() {
        return null;
    }

    @Override
    public LocalDateTime getDownloadTime() {
        return LocalDateTime.MIN;
    }
}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/room/databases/unified/entity/relations/RoomUnifiedEmbeddedCategory.kt
package com.quigglesproductions.secureimageviewer.room.databases.unified.entity.relations

import android.content.Context
import androidx.room.Embedded
import androidx.room.Ignore
import androidx.room.Junction
import androidx.room.Relation
import com.quigglesproductions.secureimageviewer.SortType
import com.quigglesproductions.secureimageviewer.checksum.FileChecksum
import com.quigglesproductions.secureimageviewer.datasource.folder.IFolderDataSource
import com.quigglesproductions.secureimageviewer.datasource.folder.IFolderDataSource.FolderSourceType
import com.quigglesproductions.secureimageviewer.enums.FileGroupBy
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDatabaseFolder
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDisplayFolder
import com.quigglesproductions.secureimageviewer.room.databases.unified.UnifiedFileDatabase
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedCategory
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedFile
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedFileCategoryCrossRef
import com.quigglesproductions.secureimageviewer.ui.enhancedfolderviewer.FolderOrigin
import java.io.File
import java.net.URL
import java.time.LocalDateTime

class RoomUnifiedEmbeddedCategory : IDisplayFolder, IDatabaseFolder {
    @Embedded
    var category: RoomUnifiedCategory? = null

    @Relation(
        parentColumn = "CategoryId", entityColumn = "FileId", associateBy = Junction(
            RoomUnifiedFileCategoryCrossRef::class
        ), entity = RoomUnifiedFile::class
    )
    var filesWithCategory: List<RoomUnifiedEmbeddedFile>? = null

    //@Relation(parentColumn = "OnlineThumbnailId",entityColumn = "OnlineId",entity = RoomDatabaseFile.class)
    @Ignore
    var thumbnailFile: RoomUnifiedEmbeddedFile? = null

    @Ignore
    private var dataSource: IFolderDataSource? = null
    override fun getDataSource(): IFolderDataSource {
        if(dataSource == null)
            dataSource = object : IFolderDataSource{
                override fun getFolderURL(): URL? {
                    return null
                }

                /*override fun getFilesFromDataSource(
                    context: Context,
                    callback: IFolderDataSource.FolderDataSourceCallback,
                    sortType: SortType
                ) {
                    callback.FolderFilesRetrieved(filesWithCategory, null);
                }*/

                override suspend fun getThumbnailFromDataSourceSuspend(
                    context: Context,
                    database: UnifiedFileDatabase
                ): Any? {
                    if(filesWithCategory!!.isNotEmpty()) {
                        val file = database.fileDao().getByOnlineId(filesWithCategory!![0].onlineId)
                        return file?.thumbnailFile
                    }
                    else
                        return null
                }

            }
        return dataSource!!
    }

    override fun hasUpdates(): Boolean {
        return false
    }

    override fun getName(): String {
        return category!!.getName()
    }

    override fun getIsAvailable(): Boolean {
        return true
    }

    override fun getUid(): Long {
        return category!!.getUid()
    }

    override fun setHasUpdates(b: Boolean) {}
    override fun getOnlineId(): Long {
        return category!!.getOnlineId()
    }

    override fun setDataSource(retrofitFolderDataSource: IFolderDataSource) {
        dataSource = retrofitFolderDataSource
    }

    override fun getFolderOrigin(): FolderOrigin {
        return FolderOrigin.ROOM
    }

    override fun sortFiles(newSortType: SortType) {
        when (newSortType) {
            SortType.NAME_ASC -> filesWithCategory!!.sortedBy { obj: IDisplayFile -> obj.name }
            SortType.NAME_DESC -> filesWithCategory!!.sortedByDescending { obj: IDisplayFile -> obj.name }
            SortType.NEWEST_FIRST -> filesWithCategory!!.sortedByDescending { obj: IDisplayFile -> obj.getDefaultSortTime() }
            SortType.OLDEST_FIRST -> filesWithCategory!!.sortedBy { obj: IDisplayFile -> obj.getDefaultSortTime() }
        }
    }

    override fun getFileGroupingType(): FileGroupBy {
        return FileGroupBy.CATEGORIES
    }

    override fun getId(): Long {
        return category!!.getUid()
    }

    override fun getThumbnailFile(): File? {
        return null
    }

    override fun getDownloadTime(): LocalDateTime {
        return LocalDateTime.MIN
    }

    override fun getSourceType(): FolderSourceType {
        return category!!.sourceType
    }

    override fun setIsAvailableOffline(value: Boolean) {
        category!!.isAvailableOffline = value
    }

    override fun getIsAvailableOffline(): Boolean {
        return category!!.isAvailableOffline
    }

    override fun isAvailableOfflineSet(): Boolean {
        return category!!.isAvailableOfflineSet
    }

    override fun getThumbnailChecksum(): FileChecksum? {
        return null
    }
}
